master/buildbot/reporters/notifier.py

```diff
@@ -0,0 +1,304 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from __future__ import absolute_import
+from __future__ import print_function
+from future.utils import string_types
+
+import abc
+
+from twisted.internet import defer
+
+from buildbot import config
+from buildbot.process.results import CANCELLED
+from buildbot.process.results import EXCEPTION
+from buildbot.process.results import FAILURE
+from buildbot.process.results import SUCCESS
+from buildbot.process.results import WARNINGS
+from buildbot.reporters import utils
+from buildbot.reporters.message import MessageFormatter as DefaultMessageFormatter
+from buildbot.reporters.message import MessageFormatterMissingWorker
+from buildbot.util import service
+
+ENCODING = 'utf8'
+
+
+class NotifierBase(service.BuildbotService):
+
+    __meta__ = abc.ABCMeta
+
+    possible_modes = ("change", "failing", "passing", "problem", "warnings",
+                      "exception", "cancelled")
+
+    def computeShortcutModes(self, mode):
+        if isinstance(mode, string_types):
+            if mode == "all":
+                mode = ("failing", "passing", "warnings",
+                        "exception", "cancelled")
+            elif mode == "warnings":
+                mode = ("failing", "warnings")
+            else:
+                mode = (mode,)
+        return mode
+
+    def checkConfig(self, mode=("failing", "passing", "warnings"),
+                    tags=None, builders=None,
+                    buildSetSummary=False, messageFormatter=None,
+                    subject="Buildbot %(result)s in %(title)s on %(builder)s",
+                    addLogs=False, addPatch=False,
+                    name=None, schedulers=None, branches=None,
+                    watchedWorkers=None, messageFormatterMissingWorker=None):
+
+        for m in self.computeShortcutModes(mode):
+            if m not in self.possible_modes:
+                if m == "all":
+                    config.error(
+                        "mode 'all' is not valid in an iterator and must be passed in as a separate string")
+                else:
+                    config.error(
```

Comment:
Do you have https://codecov.io/ chrome plugin installed?

There are a trivial small tests that are missing here to get 100% diff coverage